python/ray/spark/__init__.py

```diff
@@ -0,0 +1,593 @@
+)
+
+if not sys.platform.startswith("linux"):
+    raise RuntimeError("Ray on spark ony supports linux system.")
+
+_logger = logging.getLogger("ray.spark")
+
+_spark_dependency_error = "ray.spark module requires pyspark >= 3.3"
+try:
+    import pyspark
+
+    if Version(pyspark.__version__) < Version("3.3"):
+        raise RuntimeError(_spark_dependency_error)
+except ImportError:
+    raise RuntimeError(_spark_dependency_error)
+
+
+def wait_ray_node_available(hostname, port, timeout, error_on_failure):
+    # Wait Ray head node spin up.
+    for _ in range(timeout):
+        time.sleep(1)
+        if check_port_open(hostname, port):
+            break
+
+    if not check_port_open(hostname, port):
+        raise RuntimeError(error_on_failure)
+
+
+class RayClusterOnSpark:
+    """
+    This class is the type of instance returned by the `init_cluster` API.
+    Its main functionality is to:
+    Connect to, disconnect from, and shutdown the Ray cluster running on Apache Spark.
+    Serve as a Python context manager for the `RayClusterOnSpark` instance.
+
+    Args
+        address: The url for the ray head node (defined as the hostname and unused port on
+            Spark driver node)
+        head_proc: Ray head process
+        spark_job_group_id: The Spark job id for a submitted ray job
+        num_ray_workers: The number of workers in the ray cluster.
+    """
+
+    def __init__(self, address, head_proc, spark_job_group_id, num_ray_workers):
+        self.address = address
+        self.head_proc = head_proc
+        self.spark_job_group_id = spark_job_group_id
+        self.ray_context = None
+        self.is_shutdown = False
+        self.num_ray_workers = num_ray_workers
+
+    def _cancel_background_spark_job(self):
+        get_spark_session().sparkContext.cancelJobGroup(self.spark_job_group_id)
+
+    def connect(self):
+        import ray
+
+        if self.is_shutdown:
+            raise RuntimeError(
+                "The ray cluster has been shut down or it failed to start."
+            )
+        if self.ray_context is None:
+            # connect to the ray cluster.
+            self.ray_context = ray.init(address=self.address)
+
+            last_alive_worker_count = 0
+            last_progress_move_time = time.time()
+            while True:
+                time.sleep(10)
+                cur_alive_worker_count = len(
+                    [
+                        node
+                        for node in ray.nodes()
+                        if node["Alive"] and node["Resources"].get("CPU", 0) > 0
+                    ]
+                )
+                if cur_alive_worker_count == self.num_ray_workers:
+                    return
+
+                if cur_alive_worker_count > last_alive_worker_count:
+                    last_alive_worker_count = cur_alive_worker_count
+                    last_progress_move_time = time.time()
+                    _logger.warning(
```

Comment:
this should be an INFO?

Reply:
this is important info I hope user can always see it. Usually the log level might be warning.

Reply:
is it databricks/spark practice? Normally the log level is INFO so users should be able to see INFO logs.

I feel log level is not decided by the importance like we shouldn't log ERROR or FATAL because an INFO log is super important.

Reply:
ok let's use info level.

Reply:
changed to `info` logging